hibernate-core/src/main/java/org/hibernate/loader/ast/internal/CompoundNaturalIdLoader.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Hibernate, Relational Persistence for Idiomatic Java
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or http://www.gnu.org/licenses/lgpl-2.1.html
+ */
+package org.hibernate.loader.ast.internal;
+
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+
+import org.hibernate.engine.spi.SharedSessionContractImplementor;
+import org.hibernate.loader.NaturalIdPostLoadListener;
+import org.hibernate.loader.NaturalIdPreLoadListener;
+import org.hibernate.metamodel.mapping.EntityMappingType;
+import org.hibernate.metamodel.mapping.SingularAttributeMapping;
+import org.hibernate.metamodel.mapping.internal.CompoundNaturalIdMapping;
+import org.hibernate.metamodel.mapping.internal.MappingModelCreationProcess;
+import org.hibernate.sql.ast.Clause;
+import org.hibernate.sql.ast.tree.expression.JdbcParameter;
+import org.hibernate.sql.exec.internal.JdbcParameterBindingImpl;
+import org.hibernate.sql.exec.spi.JdbcParameterBindings;
+
+/**
+ * NaturalIdLoader implementation for compound natural-ids
+ */
+public class CompoundNaturalIdLoader<T> extends AbstractNaturalIdLoader<T> {
+
+	public CompoundNaturalIdLoader(
+			CompoundNaturalIdMapping naturalIdMapping,
+			NaturalIdPreLoadListener preLoadListener,
+			NaturalIdPostLoadListener postLoadListener,
+			EntityMappingType entityDescriptor,
+			MappingModelCreationProcess creationProcess) {
+		super( naturalIdMapping, preLoadListener, postLoadListener, entityDescriptor, creationProcess );
+	}
+
+	@Override
+	protected Object resolveNaturalIdBindValue(Object naturalIdValue, SharedSessionContractImplementor session) {
+		// the "real" form as an array, although we also accept here a Map and reduce it to
+		// the appropriately ordered array
+		if ( naturalIdValue instanceof Object[] ) {
```

Comment:
If `naturalIdValue` is of primitive type array, the above statement would return false. If that were not the original intention, we can use the following statement instead:
```
if ( naturalIdValue.getClass().isArray() ) {
... ...
}
```

Reply:
But I guess this PR assumed that primitive nutural id array has gone through the transformation to Object[] already.